accessibility/README.md

```diff
@@ -0,0 +1,495 @@
+* In some cases, a site's design may show inputs without visible labels. Sometimes this may be because other elements on the page give the input some context - for example a search text input placed directly next to a "Search" button.
+  * If a input does not have a visible label, a `<label>` element should still be added for the benefit of screen readers
+  * A label can be hidden visually by adding the `visually-hidden` class to the `label`
+  ```html
+  <!-- Label will be available to assistive technology, but visually hidden -->
+  <label class="visually-hidden" for="search">
+    Search the site
+  </label>
+
+  <input id="search" type="text" name="search">
+  ```
+
+### Ensure interactive elements have accessible text
+
+#### Don't
+
+```vue
+<a class="site-logo" href="/">
+  <brand-logo />
+</a>
+
+<button
+  type="button"
+  @click="openMenuDrawer"
+>
+  <icon-menu />
+</button>
+
+<a href="/account">
+  <icon-account />
+</a>
+
+<a href="/sale">
+  <!-- Image containing embedded text about a 10% off sale -->
+  <img src="10-off-promo-image.jpg">
+</a>
+
+
+```
+
+#### Do
+
+```vue
+<a class="site-logo" href="/">
+  <brand-logo />
+  
+  <span class="visually-hidden">
+    My Shop Name
+  </span>
+</a>
+
+<button
+  type="button"
+  @click="openMenuDrawer"
+>
+  <icon-menu />
+  
+  <span class="visually-hidden">
+    Toggle Menu
+  </span>
+</button>
+
+<a href="/account">
+  <icon-account />
+  
+  <span class="visually-hidden">
+    Account
+  </span>
+</a>
+
+<a href="/sale">
+  <!-- Image containing embedded text about a 10% off sale -->
+  <img src="10-off-promo-image.jpg" alt="Get 10% off">
+</a>
+```
+
+* When using `<button>` or `<a>` elements containing SVG elements, ensure that accessible text is also included
+  * If the text should not be visible, the `visually-hidden` class can be used
+* If a `<button>` or `<a>` contains a `<img>` element - ensure that there is either appropriate `alt` text on the image, or add a supporting `visually-hidden` span with appropriate text.
+  * Avoid using both `alt` and `visually-hidden` text for the same element, otherwise screen readers will read out possibly duplicate or irrelevant content.
+  * Setting an empty `alt` text is fine when using `visually-hidden` text - see [Alt Text](#alt-text) for more details.
+
+### Ensure correct usage of `aria-labelledby` and `aria-label` 
+
+* The `aria-labelledby` and `aria-label` attributes provide another method for adding labels to elements
+* These are alternatives to using a `visually-hidden` element showcased in the examples above
+
+#### `aria-labelledby`
+
+* `aria-labelledby` can be used to label an element based on the text content of another existing element by it's unique `id`
+* If a heading is present inside a region, this could be a good candidate for `aria-labelledby`
+
+```html
+<section aria-labelledby="template-title">
```

Comment:
Might want to flag why `section` elements should be labelled in the first place.

I'm still doing it because I think aXe once flagged that all `section` elements should be labelled.

Reply:
Should we consider removing `aria-labelledby` on all our `section` elements?

I have been looking through the documentation that @georgia-wmw linked, along with adding some further reading and testing and I can't see that labelling all `section` elements is _required_ - it seems more like the opposite, that `section` elements should only be labelled in certain circumstances:

From the [Scott O'Hara blog post](https://www.scottohara.me/blog/2021/07/16/section.html#use-with-purpose):
> Regarding the `<section>` element’s implicit ARIA role, it is best named – thus exposing its role – only when absolutely necessary. Overpopulating a web page with landmarks will reduce their ability to help users find the most important parts of web pages. In other words, if everything’s a landmark then there’s nothing really special about them, is there? Instead, rely on proper heading levels to indicate implicit sections and sub-sections of your content.  

From [MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/region_role#accessibility_concerns):
> Use sparingly! [Landmark roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles#3._landmark_roles) are intended to be used sparingly, to identify larger overall sections of the document. Using too many landmark roles can create "noise" in screen readers, making it difficult to understand the overall layout of the page.

I also did a basic test of some `<section>` elements without `aria-labelledby` on this codepen - https://codepen.io/jamespeilow/pen/vYvEeXN 
I did not see any accessibility issues flagged by Lighthouse or aXe Dev Tools, nor could I find any related rules in the [aXe Dev tools List of Rules](https://dequeuniversity.com/rules/axe/4.7)

I also did some testing with Voiceover on Mac on some of our recently launched Canvas sites, and there's a good amount of sites where the element referenced by `aria-labelledby` does not exist.
Given that the default ID for the `aria-labelledby` attribute that the Canvas `component` command creates contains the `section.id` - this results in Voiceover reading out the `id` string - for example: `section-title-template--16778842603746__5f8cbe58-204b-4065-b774-63f69e728aaf`. (Seen on [Stephen Webster](https://www.stephenwebster.com/))

If using `section` with `aria-labelledby` is best practice, I'm happy to keep this in Canvas, update this documentation and work on developer training to make sure its used properly and not ignored - but I couldn't find anything to back up using it on every section.

Also aware this could be broken out to a different discussion/ticket as it affects core Canvas.

**tldr: It looks like it might not be recommended to have `aria-labelledby`/`aria-label` on every `section` element, so it might be worth reviewing our usage of it**

Reply:
Stephen Webster is like a weird fluke of a site where its lighthouse score is/was 99-100 but when you actually interact using keyboard-only/a screenreader it becomes ~~a nightmarish hellscape~~ tricky to navigate 😵‍💫 Anyways

You're correct in that too many landmarks can get confusing for a webpage. I did some digging and [technique ARIA13](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA13) recommends using `aria-labelledby` for landmarks & regions, but then [technique ARIA20](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA20) states 'don't use too many landmarks'. So the instructions seem a bit out of order there. [MDN also states to use `region`s with labels sparingly](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/region_role#accessibility_concerns) so that aXe flag could have been deprecated.

We could try and control landmark labels from CANVAS instead of passing this to every component. What if we only explicitly label PDPs and PLPs? I think those are pretty standard and predictable parts of all/most client websites. Then we can rely on appropriate HTML headings for segmenting the content itself (conscious that this is often up to client/designs and not always something in our control). The `aria-labelledby`s could use `{{ product.title }}` and `{{ collection.title }}` by default too, and there should be a loading state for both of these elements, so the ID would be present straight from page load (though that might get a bit repetitive)

Reply:
Thanks for that ARIA13 link @georgia-wmw.
The one thing that isn't clear with that is its relation to the `<section>` element - ARIA13 describes that "landmarks and regions" should use `aria-labelledby` - However a `<section>` element seems like a bit of a special case in that it only becomes a `region`/gets an implied `role="region"` when it is given an accessible name:

[From MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/region_role#description)
> Using the [<section>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section) element will automatically communicate a section has a role of region if it is given an accessible name. Developers should always prefer using the correct semantic HTML element, in this case <section>, over using ARIA.

I do think we could consider still using `aria-labelledby` for some of the key default Canvas components, but then removing it from the component template for other section type components - a developer could always add it back to a certain section if needed.

Loading state is probably another thing to consider - If a section is loading and the `aria-labelledby` element depends on data that isn't ready, I wonder if having "loading" as the `aria-labelledby`text is better or if its better to have nothing, or some other way of handling that

Reply:
Do we just keep the labels on the sections that use the `template-title` ID instead of the `section-title-{{ section.id }}`? The assumption being those are on the 'main' components which are important to the page and not used on section everywhere sections.

Reply:
I think that's a good idea! It's an easily identifiable landmark & most/all instances of `template-title` on the canvas repo reference a `h1` so there should only be 1 per page (keeping them sparse)

Reply:
I agree - Keeping the labels just for `template-title` ID sounds like a good approach considering they are the "main" sections/components on the site.

Reply:
@jamespeilow Excellent, I'll look at updating the `component` command and removing labels from non-template ID sections, do you want to flag which sections should be labelled?